### DISCUSSION TITLE
lazy.nvim `optional` attribute

### DISCUSSION THREAD

**Discussion by dpetka2001** [LazyVim/LazyVim#1049] [General]
First discussed in #640, the `optional` attribute made its appearance. I thought I understood the concept of it in that PR, as both `mini.starter` and `alpha.nvim` would only come into the `project` spec if they were first defined somewhere else in the personal user spec (according to what the user decides to use). But then I also saw `optional` in https://github.com/LazyVim/LazyVim/blob/ddb819cb9ae3a9b573be1f7d57c537dfa8399c41/lua/lazyvim/plugins/extras/lang/rust.lua#L33-L36 and https://github.com/LazyVim/LazyVim/blob/ddb819cb9ae3a9b573be1f7d57c537dfa8399c41/lua/lazyvim/plugins/extras/lang/clangd.lua#L101-L108 In both of them I don't understand why `mason` should be `optional`. Isn't there a `mason` spec already configured? What is the point of it being declared as `optional`? And in the `clangd` extras spec it's also declared as a dependency to `nvim-dap` which is already declared as `optional`. So if a user doesn't have a debugging spec enabled, that whole `nvim-dap` part won't make it into the final spec. Doesn't that also mean that `mason` won't get invoked as it's a dependency of `nvim-dap`. Why does it have to be `optional`? Excuse my ignorance, but I'm genuinely perplexed as to `optional`'s usage in the last 2 extras specs. If someone could shed some light as to its correct usage, so that I can also have a clear aspect of its usage in my mind, I would be really grateful.

**Comment by folke** [comment]
Users can disable `mason` in their config, so in those cases `optional` would come into effect. Although I'm not 100% sure that would work as expected :)

**Reply by dpetka2001** [reply]
I understand what you're saying. It just seems baffling to me, as I thought that `mason` is one of the base plugins (although "base" can be up to user interpretation). Because for me as a user, that would mean that it would be good practice to configure most of the plugins as `optional` in this case. I understand that `optional` is to be mainly used by Neovim distros (according to lazy.nvim docs), but still its usage (regarding mason specifically) looked confusing to me in the last 2 PRs that I mentioned. Anyhow, thanks for clarifying @folke. I understand what you were trying to do and it looks clearer to me now.

**Reply by folke** [reply]
I do agree it doesn't make a lot of sense to make mason optional, since it should be indeed a core plugin. But at the same time it also doesn't hurt
